File: skills/hydric-gateway-api-user/examples/search-pools.ts
/**
 * Hydric Gateway API - Pool Search Example (TypeScript)
 *
 * This example demonstrates how to search for liquidity pools
 * using the Hydric Gateway API.
 *
 * Usage:
 *   1. Replace YOUR_API_KEY with your actual API key
 *   2. Run: npx ts-node search-pools.ts
 */

const API_KEY = 'YOUR_API_KEY';
const BASE_URL = 'https://api.hydric.org/v1';

// Chain IDs for reference
const ChainId = {
  ETHEREUM: 1,
  BASE: 8453,
  SCROLL: 534352,
  MONAD: 143,
  UNICHAIN: 130,
  HYPER_EVM: 999,
  PLASMA: 9745,
} as const;

// Well-known token addresses
const TOKENS = {
  ETH: '0x0000000000000000000000000000000000000000',
  WETH_BASE: '0x4200000000000000000000000000000000000006',
  USDC_ETH: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
  USDC_BASE: '0x833589fcd6edb6e08f4c7c32d4f71b54bda02913',
};

interface BlockchainAddress {
  chainId: number;
  address: string;
}

interface SearchPoolsRequest {
  tokensA: BlockchainAddress[];
  tokensB?: BlockchainAddress[];
  filters?: {
    minimumTotalValueLockedUsd?: number;
    blockedPoolTypes?: string[];
    blockedProtocols?: string[];
  };
  config?: {
    limit?: number;
    orderBy?: {
      field: 'tvl' | 'yield';
      direction: 'asc' | 'desc';
      timeframe?: '24h' | '7d' | '30d' | '90d';
    };
    cursor?: string;
  };
}

async function searchPools(request: SearchPoolsRequest) {
  const response = await fetch(`${BASE_URL}/pools/search`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${API_KEY}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(`API Error: ${error.error?.message || response.statusText}`);
  }

  const json = await response.json();
  return json.data;
}

// Example 1: Find all pools containing ETH on Ethereum
async function findEthPools() {
  console.log('=== Finding all ETH pools on Ethereum ===\n');

  const result = await searchPools({
    tokensA: [{ chainId: ChainId.ETHEREUM, address: TOKENS.ETH }],
    config: {
      limit: 5,
      orderBy: { field: 'tvl', direction: 'desc' },
    },
    filters: {
      minimumTotalValueLockedUsd: 1000000,
    },
  });

  console.log(`Found ${result.pools.length} pools:\n`);
  for (const pool of result.pools) {
    const tokens = pool.tokens.map((t: any) => t.symbol).join('-');
    console.log(`  ${pool.protocol.name} | ${tokens}`);
    console.log(`    TVL: $${pool.balance.totalValueLockedUsd.toLocaleString()}`);
    console.log(`    24h Yield: ${pool.stats.stats24h.yield.toFixed(2)}%`);
    console.log(`    Address: ${pool.address}\n`);
  }
}

// Example 2: Find ETH-USDC pair pools on Base sorted by yield
async function findBestYieldEthUsdcOnBase() {
  console.log('=== Best ETH-USDC pools on Base by 24h yield ===\n');

  const result = await searchPools({
    tokensA: [{ chainId: ChainId.BASE, address: TOKENS.WETH_BASE }],
    tokensB: [{ chainId: ChainId.BASE, address: TOKENS.USDC_BASE }],
    config: {
      limit: 5,
      orderBy: { field: 'yield', direction: 'desc', timeframe: '24h' },
    },
    filters: {
      minimumTotalValueLockedUsd: 50000,
    },
  });

  console.log(`Found ${result.pools.length} ETH-USDC pools:\n`);
  for (const pool of result.pools) {
    console.log(`  ${pool.protocol.name} (${pool.type})`);
    console.log(`    TVL: $${pool.balance.totalValueLockedUsd.toLocaleString()}`);
    console.log(`    24h Yield: ${pool.stats.stats24h.yield.toFixed(2)}%`);
    console.log(`    7d Yield: ${pool.stats.stats7d.yield.toFixed(2)}%`);
    console.log(`    Fee: ${pool.feeTier.feeTierPercentage}%\n`);
  }
}

// Example 3: Paginate through all results
async function paginateThroughPools() {
  console.log('=== Paginating through all ETH pools ===\n');

  let cursor: string | null = null;
  let page = 1;
  let totalPools = 0;

  do {
    const result = await searchPools({
      tokensA: [{ chainId: ChainId.ETHEREUM, address: TOKENS.ETH }],
      config: {
        limit: 10,
        cursor: cursor || undefined,
      },
    });

    totalPools += result.pools.length;
    console.log(`Page ${page}: ${result.pools.length} pools`);
    cursor = result.nextCursor;
    page++;

    // Stop after 3 pages for this demo
    if (page > 3) break;
  } while (cursor);

  console.log(`\nTotal pools fetched: ${totalPools}`);
}

// Run all examples
async function main() {
  try {
    await findEthPools();
    console.log('\n' + '='.repeat(50) + '\n');
    await findBestYieldEthUsdcOnBase();
    console.log('\n' + '='.repeat(50) + '\n');
    await paginateThroughPools();
  } catch (error) {
    console.error('Error:', error);
  }
}

main();

File: skills/hydric-gateway-api-user/hydric-gateway-api-user/examples/search_pools.py
"""
Hydric Gateway API - Pool Search Example (Python)

This example demonstrates how to search for liquidity pools
using the Hydric Gateway API.

Usage:
    1. Replace YOUR_API_KEY with your actual API key
    2. Install requests: pip install requests
    3. Run: python search_pools.py
"""

import requests
from typing import Optional

API_KEY = "YOUR_API_KEY"
BASE_URL = "https://api.hydric.org/v1"


class ChainId:
    ETHEREUM = 1
    BASE = 8453
    SCROLL = 534352
    MONAD = 143
    UNICHAIN = 130
    HYPER_EVM = 999
    PLASMA = 9745


class Tokens:
    ETH = "0x0000000000000000000000000000000000000000"
    WETH_BASE = "0x4200000000000000000000000000000000000006"
    USDC_ETH = "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"
    USDC_BASE = "0x833589fcd6edb6e08f4c7c32d4f71b54bda02913"


def search_pools(
    tokens_a: list[dict],
    tokens_b: Optional[list[dict]] = None,
    min_tvl: float = 0,
    limit: int = 10,
    order_by: str = "tvl",
    direction: str = "desc",
    timeframe: str = "24h",
    cursor: Optional[str] = None,
) -> dict:
    """Search for liquidity pools."""
    
    payload = {
        "tokensA": tokens_a,
        "config": {
            "limit": limit,
            "orderBy": {
                "field": order_by,
                "direction": direction,
                "timeframe": timeframe,
            },
        },
        "filters": {
            "minimumTotalValueLockedUsd": min_tvl,
        },
    }
    
    if tokens_b:
        payload["tokensB"] = tokens_b
    
    if cursor:
        payload["config"]["cursor"] = cursor
    
    response = requests.post(
        f"{BASE_URL}/pools/search",
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "Content-Type": "application/json",
        },
        json=payload,
    )
    
    if not response.ok:
        error = response.json()
        raise Exception(f"API Error: {error.get('error', {}).get('message', response.text)}")
    
    return response.json()["data"]


def find_eth_pools():
    """Find all pools containing ETH on Ethereum."""
    print("=== Finding all ETH pools on Ethereum ===\n")
    
    result = search_pools(
        tokens_a=[{"chainId": ChainId.ETHEREUM, "address": Tokens.ETH}],
        min_tvl=1_000_000,
        limit=5,
    )
    
    print(f"Found {len(result['pools'])} pools:\n")
    for pool in result["pools"]:
        tokens = "-".join(t["symbol"] for t in pool["tokens"])
        print(f"  {pool['protocol']['name']} | {tokens}")
        print(f"    TVL: ${pool['balance']['totalValueLockedUsd']:,.2f}")
        print(f"    24h Yield: {pool['stats']['stats24h']['yield']:.2f}%")
        print(f"    Address: {pool['address']}\n")


def find_best_yield_eth_usdc_on_base():
    """Find ETH-USDC pair pools on Base sorted by yield."""
    print("=== Best ETH-USDC pools on Base by 24h yield ===\n")
    
    result = search_pools(
        tokens_a=[{"chainId": ChainId.BASE, "address": Tokens.WETH_BASE}],
        tokens_b=[{"chainId": ChainId.BASE, "address": Tokens.USDC_BASE}],
        min_tvl=50_000,
        limit=5,
        order_by="yield",
        timeframe="24h",
    )
    
    print(f"Found {len(result['pools'])} ETH-USDC pools:\n")
    for pool in result["pools"]:
        print(f"  {pool['protocol']['name']} ({pool['type']})")
        print(f"    TVL: ${pool['balance']['totalValueLockedUsd']:,.2f}")
        print(f"    24h Yield: {pool['stats']['stats24h']['yield']:.2f}%")
        print(f"    7d Yield: {pool['stats']['stats7d']['yield']:.2f}%")
        print(f"    Fee: {pool['feeTier']['feeTierPercentage']}%\n")


def get_token_price(chain_id: int, token_address: str) -> float:
    """Get the USD price of a token."""
    response = requests.get(
        f"{BASE_URL}/tokens/prices/{chain_id}/{token_address}/usd",
        headers={"Authorization": f"Bearer {API_KEY}"},
    )
    
    if not response.ok:
        raise Exception(f"Failed to get price: {response.text}")
    
    return response.json()["data"]["price"]


def calculate_portfolio_value():
    """Calculate the value of a sample portfolio."""
    print("=== Calculating Portfolio Value ===\n")
    
    portfolio = [
        {"chainId": ChainId.ETHEREUM, "address": Tokens.ETH, "amount": 2, "symbol": "ETH"},
        {"chainId": ChainId.ETHEREUM, "address": Tokens.USDC_ETH, "amount": 1000, "symbol": "USDC"},
    ]
    
    total_value = 0
    for token in portfolio:
        price = get_token_price(token["chainId"], token["address"])
        value = price * token["amount"]
        total_value += value
        print(f"  {token['amount']} {token['symbol']}: ${value:,.2f} (@ ${price:,.2f})")
    
    print(f"\n  Total Portfolio Value: ${total_value:,.2f}")


if __name__ == "__main__":
    try:
        find_eth_pools()
        print("\n" + "=" * 50 + "\n")
        find_best_yield_eth_usdc_on_base()
        print("\n" + "=" * 50 + "\n")
        calculate_portfolio_value()
    except Exception as e:
        print(f"Error: {e}")
